OracleGoldenGate/21c/bin/deployment-main.sh

```diff
@@ -121,6 +124,37 @@ function setup_deployment_directories() {
     chown -R ogg:ogg {} \;
 }
 
+##
+##  r u n _ p r e s t a r t _ c o m m a n d
+##
+## Hook for launching custom command in the container prior to ogg start
+## If defined, the command identified by ${RUN_COMMAND} will be run.
+## 
+## The value of ${RUN_COMMAND_BACKGROUND} by default is set to true.
+##     When ${RUN_COMMAND_BACKGROUND} is true:
+##          The command identified by ${RUN_COMMAND} will be run in the background 
+##          and its return status is ignored
+##     When ${RUN_COMMAND_BACKGROUND} is false:
+##          The command identified by ${RUN_COMMAND} will be run in the foreground 
+##          and must return a status of 0 in order to proceed with the start sequence
+##
+function run_prestart_command {
```

Comment:
I think this function could consist of a single one-line statement:

```sh
eval "$(RUN_COMMAND}" || abort "${RUN_COMMAND} did not complete successfully!"
```

An empty value for `RUN_COMMAND` becomes a no-op.

If `RUN_COMMAND` is terminated with `&`, the task is executed as a background job. That would also allow for more complex background tasks run with `nohup` and similar wrappers.

Reply:
The mounted volume pattern is nice. I think this is the route I will approach as it gives a known set of tested scripts that are run.

I was not a great fan of the arbitrary RUN_COMMAND approach to begin with, due to the ability to run any arbitrary code in the container as a privileged user.  I avoided the eval statement even more so, due to it's expansion allowing for even more complex execution patterns.

The directory/mounted volume approach appears to force a foreground linear lexicographical approach.

Is there any desire to allow for asynchronous background processes to be started as well?

This could be achieved with an intuitive directory structure:

| - hooks
&nbsp;&nbsp;&nbsp;&nbsp;| -- background
&nbsp;&nbsp;&nbsp;&nbsp;| -- prestart

Although possible, I did not propose an exit/cleanup folder.  This enhancement could be augmented in the future, but the current signal handling and health monitoring would need a good bit of rework to ensure OGG was in the correct state while not having the container terminated/restarted by the container management system prior to the scripts finishing.

Additional feedback would be appreciate before proceeding.

Thanks again.

Reply:
Updated to use the directory pattern ${OGG_HOME}/scripts/setup and ${OGG_HOME}/scripts/startup

The backgrounding issue can be handled on a per script basis inside the script itself, rather than as an arbitrary flag in the config.  Allowing user determination on what runs in the background and in what order.

The scripts will be run lexicographically and will recursively span subdirectories.
The order preference of files is determined only by the lexicographical order on a per directory basis.
ie: 
> 02-sanity.sh
> 01-connectivity-validation\
>>> 02-test-hostA.sh 
>>>01-test-hostZ.sh 

will be run in the lexicographic order:
1. 01-test-hostZ.sh
2. 02-test-hostA.sh
3. 02-sanity.sh

SQLPlus is dynamically located and .sql files can be run.
Scripts in */setup will be run prior to any other startup sequence
Scripts in */startup will be run after ogg/nginx successfully start